ua-snap/cmip6-utils
# Add QC

This PR adds a `indicators/qc.py` script and additional lookup tables in `indicators/luts.py` to the `indicators` branch.  In order to test this PR using Prefect, you will need to use the `add_qc_task` branch from the [prefect](https://github.com/ua-snap/prefect/tree/add_qc_task) repo and run the version of `indicators/generate_indicators.py` script found in that branch. That flow now has modifications that will call the `indicators/qc.py` script from this repo after downloading. This PR also removes the earlier validation functions from `indicators/indicators.py` - these are replaced by the QC functions in `indicators/qc.py`. 

The `indicators/slurm.py` is also modified to remove references to the main output directory (`SCRATCH_DIR`) and adds hard-coded values for `sbatch_head_kwargs` and `indicators_script`. These were originally pulled from `indicators/config.py` but since the paths were being built from environment variables which were not found during the Prefect deployment, they defaulted to `/beegfs/CMIP6/snapdata/...` directories. To avoid overwriting files in that shared directory, I hard-coded the previously mentioned variables for testing. Its a temporary solution until we think up something more clever!

Similarly, the `add_qc_task` branch from the prefect repo contains a hard-coded definition of `conda_init_script` in `indicators/indicator_functions.py`. The parameters I used for a successful flow run of `indicators/generate_indicators.py` from that branch are below:

```
{
  "ssh_username": "jdpaul3",
  "ssh_private_key_path": "/Users/joshpaul/.ssh/id_rsa",
  "branch_name": "add_qc",
  "working_directory": "/import/beegfs/CMIP6/jdpaul3/scratch",
  "indicators": "rx1day",
  "models": "CESM2 GFDL-ESM4 TaiESM1",
  "scenarios": "historical ssp126 ssp245 ssp370 ssp585",
  "slurm_script": "/import/beegfs/CMIP6/jdpaul3/scratch/cmip6-utils/indicators/slurm.py",
  "qc_script": "/import/beegfs/CMIP6/jdpaul3/scratch/cmip6-utils/indicators/qc.py",
  "input_dir": "/import/beegfs/CMIP6/arctic-cmip6/regrid/",
  "output_dir": "/import/beegfs/CMIP6/jdpaul3/scratch/"
}
```

The resulting tree built in the `{output_dir}` should wind up looking like this:
```
.
├── cmip6-utils
│   ├── archive
│   ├── indicators
│   │   └── __pycache__
│   ├── regridding
│   │   └── tests
│   └── transfers
│       ├── batch_files
│       └── tests
├── output
│   ├── CESM2
│   │   ├── historical
│   │   │   └── rx1day
│   │   ├── ssp126
│   │   │   └── rx1day
│   │   ├── ssp245
│   │   │   └── rx1day
│   │   ├── ssp370
│   │   │   └── rx1day
│   │   └── ssp585
│   │       └── rx1day
│   ├── GFDL-ESM4
│   │   ├── historical
│   │   │   └── rx1day
│   │   ├── ssp126
│   │   │   └── rx1day
│   │   ├── ssp245
│   │   │   └── rx1day
│   │   ├── ssp370
│   │   │   └── rx1day
│   │   └── ssp585
│   │       └── rx1day
│   └── TaiESM1
│       ├── historical
│       │   └── rx1day
│       ├── ssp126
│       │   └── rx1day
│       ├── ssp245
│       │   └── rx1day
│       ├── ssp370
│       │   └── rx1day
│       └── ssp585
│           └── rx1day
├── qc
└── slurm

```

**TO TEST:**

- See if you can successfully generate the indicators and QC products. 
- Check output logs in the Prefect dashboard (or in the terminal) to make sure you are getting QC messages indicating number of errors found (if any). They should look like this: 

```
13:10:05.565 | INFO    | Flow run 'full_qc_test' - Created task run 'qc-0' for task 'qc'
13:10:05.565 | INFO    | Flow run 'full_qc_test' - Executing 'qc-0' immediately...
13:10:09.625 | INFO    | Task run 'qc-0' - QC process started...
13:10:09.626 | INFO    | Task run 'qc-0' - QC process complete: 0 errors found. See /import/beegfs/CMIP6/jdpaul3/scratch/qc/qc_error.txt for error log.
13:10:09.628 | INFO    | Task run 'qc-0' - QC script run successfully
13:10:09.666 | INFO    | Task run 'qc-0' - Finished in state Completed()
```

- If indicators were successfully generated, and no QC errors reported, go mess around with some files (rename them, delete them, erase the output of the slurm job `*.out` files, etc...) and then run the `indicators/qc.py` from your terminal and make sure you are seeing error messages that correspond to the files you bugged up. Something like:
```

(cmip6-utils) [jdpaul3@chinook04 indicators]$ python qc.py --out_dir /beegfs/CMIP6/jdpaul3/scratch
QC process started...
QC process complete: 1 errors found. See /beegfs/CMIP6/jdpaul3/scratch/qc/qc_error.txt for error log.

```
- View the new QC tables in `luts.py`
- Suggest more QC tasks if you think of any!
